File: pingdart-dart/lib/src/services/ai_service.dart
import 'package:http/http.dart' as http;
import 'dart:convert';

class AiService {
  final http.Client client;
  final String apiKey;
  final String baseUrl;

  AiService({required this.client, required this.apiKey, required this.baseUrl});

  Future<String> callAiApi(String message, {Function(String)? onProgress, String model = 'chinnuai:1.1', Map<String, dynamic>? options}) async {
    final payload = {
      'message': message,
      'stream': true,
      'model': model,
      if (options != null) ...options,
    };

    final request = http.Request('POST', Uri.parse('${baseUrl}ai/chinuai-chat'))
      ..headers.addAll({
        'x-api-key': apiKey,
        'Content-Type': 'application/json',
      })
      ..body = jsonEncode(payload);

    final response = await client.send(request);
    
    var fullResult = "";
    await response.stream.transform(utf8.decoder).listen((line) {
      if (line.startsWith('data: ')) {
        final dataStr = line.substring(6).trim();
        if (dataStr.isEmpty) return;
        try {
          final parsed = jsonDecode(dataStr);
          if (parsed['chunk'] != null) {
            fullResult += parsed['chunk'];
            if (onProgress != null) onProgress(parsed['chunk']);
          }
        } catch (e) {
          // Fallback
        }
      }
    }).asFuture();

    return fullResult;
  }
}

File: pingdart-php/src/Services/AiService.php
namespace PingDart\SDK\Services;

use GuzzleHttp\Client;

class AiService {
    /** @var Client */
    private $http;

    public function __construct(Client $http) {
        $this->http = $http;
    }

    public function callAiApi(string $message, callable $onProgress = null, string $model = 'chinnuai:1.1', array $options = []) {
        $payload = array_merge([
            'message' => $message,
            'stream' => true,
            'model' => $model
        ], $options);

        try {
            $response = $this->http->post('ai/chinuai-chat', [
                'json' => $payload,
                'stream' => true
            ]);

            $body = $response->getBody();
            $fullResult = "";

            while (!$body->eof()) {
                $line = $this->readLine($body);
                if (strpos($line, 'data: ') === 0) {
                    $jsonStr = substr($line, 6);
                    if (empty(trim($jsonStr))) continue;

                    $parsed = json_decode($jsonStr, true);
                    if (isset($parsed['chunk'])) {
                        $fullResult .= $parsed['chunk'];
                        if ($onProgress) {
                            $onProgress($parsed['chunk']);
                        }
                    }
                    if (isset($parsed['done']) && $parsed['done']) {
                        break;
                    }
                }
            }

            return $fullResult;
        } catch (\Exception $e) {
            throw new \Exception("AI API Error: " . $e->getMessage());
        }
    }

    private function readLine($stream) {
        $buffer = "";
        while (!$stream->eof()) {
            $char = $stream->read(1);
            if ($char === "\n") break;
            $buffer .= $char;
        }
        return trim($buffer);
    }
}
